ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/ActivePeers.hs

```diff
@@ -49,6 +50,9 @@ belowTarget actions
                           targetNumberOfActivePeers
                         }
             }
+  | Map.null (EstablishedPeers.establishedReady establishedPeers)
+  = GuardedSkip (Min <$> EstablishedPeers.minActivateTime establishedPeers)
```

Comment:
I guess so. This whole thing with a time delay on the re-promotion from warm to hot is rather crude.

It's just one fixed delay, when in the end what we'd want is to look at the tip sample protocol result and only consider promoting if that tells us that the node might be a good hot candidate again.

I mean it's not terrible, and it's probably ok for now, but we should think about the longer term too. Perhaps what we want is to allow the policy for picking warm peers to promote to be allowed to make no progress, if none of the current warm peers are suitable. Then we could have a policy that relies on looking at the tip sample results.

If we went that direction we'd need a governor change to let it deal with warm->hot picking policies that do not always make progress. It'd need some back-off and retry there, so it can try again once we have some more info or some new candidates.

Reply:
> This whole thing with a time delay on the re-promotion from warm to hot is rather crude.

What is progress? A few possible choices:
* _good progress_: adding  a peer from whom we have a chance of getting a new header / block fastest with some reasonable chance of success
* _fake progress_: adding a peer from whom we have little chance of getting a new header before getting it from any other peer.
* _governor progress_: progress from the governor perspective (I think this is the meaning you're referring to)

Is _fake progress_ a bad thing? Maybe not: having more hot peers allows transactions to flow faster between peers, and since SPOs are prepared that all their hot peers are _good_ (both can deliver blocks and tx-s), they will be prepared for _fake_ ones too.  But on the other side, if we organize the network based on dissemination of headers / blocks, the transactions should have similarly good propagation characteristics (they flow in different direction but using the same path as headers). So we don't need to worry about them.

 As of _governor progress_ I think we should allow no progress for the above reasons.

Technically it shouldn't be difficult to add additional metrics to `KnownPeerInfo` available in policies that can be read from an external store, it just needs an `STM` interface, but indeed we'd need to extend the governor side.  When no progress is made for a longer period, maybe it should do more aggressive churn of warm peers. But we should avoid making it too complex.

> It'd need some back-off and retry there, so it can try again once we have some more info or some new candidates.

The current implementation will retry as long as its under its targets and as soon as there are available warm peers. Are you thinking about changing the rate of governor adaptation? The information about peers will slowly change, and we want the governor to adapt at a similar scale and it could be slower than the changes in the environment (connection quality fluctuations, peers becoming busy at times, etc).